src/index.js

```diff
@@ -509,3 +516,87 @@ function sendResponse(response: $Response, type: string, data: string): void {
   response.setHeader('Content-Length', String(chunk.length));
   response.end(chunk);
 }
+
+/**
+ * Helper function to get the first word from string.
+ * 
+ * @param   { string } text - The full string.
+ * 
+ * @returns { string }      - First word.
+ */
+function sliceFirstWord(text: string): string {
+  let slicedText = text; 
+
+  const firstSpaceIndex = slicedText.indexOf(' ');
+
+  if(firstSpaceIndex !== -1) {
+    slicedText = slicedText.slice(0, firstSpaceIndex);
+  }
+
+  const firstEndRowIndex = slicedText.indexOf('\n');
+
+  if(firstEndRowIndex !== -1) {
+    slicedText = slicedText.slice(0, firstEndRowIndex);
+  }
+
+  return slicedText;
+}
+
+/**
+ * Helper recursive function that finds all the fragments from the server that are used in the current request.
+ * 
+ * @param   { string }      serverFragments - Fragments from the server.
+ * @param   { string }      query           - Query from the request.
+ * @param   { Set<string> } fragmentsInUsed - Set of relevant fragments.
+ * 
+ * @returns { Set<string> }                 - relevant fragments for current request.
+ */
+function findFragments(serverFragments: string, query: string, fragmentsInUsed: Set<string>): Set<string> {
```

Comment:
it would be much simpler and more concise to use the `graphql` `visit()` function here I think

Reply:
https://graphql.org/graphql-js/language/#visit just parse the operation string to AST, and visit all the `FragmentSpreadDefinition` node types i think it would be for this case.

```js
import { visit, print } from 'graphql'
const operationAST = parse(query)
const serverFragments  = ''
visit(operationAst, {
  FragmentDefinition(node) {
    // do matching/filtering logic here using full ast node.
    // node.name.value, etc
    serverFragments += print(node)
  }
})
```

Reply:
I could be wrong though, about which option is more performant, as parsing each query and then re-printing the nodes entails another expense, though much of this could be memoized, eventually cached even?